src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.cs

```diff
@@ -322,7 +322,23 @@ public bool Join(int millisecondsTimeout)
         /// </summary>
         internal const int OptimalMaxSpinWaitsPerSpinIteration = 64;
```

Comment:
Is 64 hardcoded value calculated for pre-Skylake or Skylake?
It's interesting that Mono uses 7 https://github.com/dotnet/runtime/blob/c59b5171e4fb2b000108bec965f8ce443cb95a12/src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs#L137-L138

Also, I wonder if we can rely on `--instruction-set:skylake` to avoid auto-detection

Reply:
I do not know. That is tracked by https://github.com/dotnet/corert/issues/4430 
I think users would typically use the default settings when compiling.

Generally if one has a fixed number of spins and thus requires calibrations, it is already an indication of arguably broken code. Spins are typically used in contention backoffs and should dynamically scale with contention. 
I guess the pattern with fixed number of spin iterations is still common enough that we have calibration that maintains timing compat and the issue tracks that.

Reply:
Here I just want to fix the assert, which is a nuisance when testing, and do a trivial change to avoid long uninterruptible spins.

Reply:
> Is 64 hardcoded value calculated for pre-Skylake or Skylake?

I think it's pre-skylake. We normalize to what it means on skylake in RhSpinWait.

> Also, I wonder if we can rely on --instruction-set:skylake to avoid auto-detection

Did AMD do a similar timing related change to their CPUs around when they started implementing skylake features? If not, we would still need this. It will also auto-tune to whatever changes future CPUs make to these instructions.

Reply:
> Did AMD do a similar timing related change to their CPUs

They kinda did 
(looking at 'pause' latency in [Agner Fog tables](https://www.agner.org/optimize/instruction_tables.pdf)):

AMD:
- Zen1: ~3  clocks
- Zen2, Zen3:  ~65  clocks

Intel
- Sady Bridge: 11 clocks
- Haswell, Broadwell: 9 clocks
- Skylake: 160 clocks

Note that this is in clocks so it will vary with CPU clock rate. It will be different for different CPU flavors. It may vary dynamically as CPU adjust for load and thermal throttling. 
It may be different for different cores in the same package.

The idea is that it will be used in adaptive spinning like exponential backoff and such, so the actual latency should not matter much. Yet it does because people hardcode spin counts.

ARM64 has similar "nop-with-a-hint" instruction ([`yield`](https://developer.arm.com/documentation/dui0473/m/arm-and-thumb-instructions/yield) ).
I do not know how long it may take.